### DISCUSSION TITLE
Async Class Constructors

### DISCUSSION THREAD

**Discussion by onebarloop** [neuefische/questions#429] [Web] [answered]
Merry Christmas / Happy Holidays everybody!

I am wondering: Is it possible to get an asnyc function to work inside a class constructor? 

```typescript
class Artist {
  artistName: string;
  location: {
    city: string;
    streetname: string;
    number: number;
  };
  slug: string;
  tattoos: string[];
  position: number[]

  constructor(
    artistName: string,
    city: string,
    streetname: string,
    number: number,
    tatoos: string[]
  ) {
    this.artistName = artistName;
    this.location = { city, streetname, number };
    this.slug = normalize(artistName);
    this.tattoos = tatoos;
    this.position = fetchGeoData(streetname, number, city)
  }
}
```

`fetchGeoData` is a function, that returns a geodata numbers-array. But since it is an asnyc function, I only get a promise back during instantiation (`new Artist()`-call). Is there a way to get async functions to work inside a constructor?

My workaround for now: Call `fetchGeoData` AFTER the object has been created, and overwrite the initial value (null) of the position-property with the actual return value. But I don't like this very much. 

[This](https://github.com/onebarloop/capstone-project/pull/22/files#diff-03f4dae20004b40c1cf621d00a700a52e85e7c1d18f15abf5807c8dc457f3aaa) is the async function, and [here](https://github.com/onebarloop/capstone-project/pull/22/files#diff-24108f480f6aeb797d97be73c9f3fdb788d1853070a78ae4a9b884cb55977562) it is called after the creation of a new Artist-object. The Artist Class looks like [this](https://github.com/onebarloop/capstone-project/pull/22/files#diff-258f5a0e7955f2a5eb4d888e377323bbb3de7ed836ba6c5d3a57745de1e5e984) for now.

I tried to implement this solution, but I can't get it to work: https://stackoverflow.com/questions/43431550/async-await-class-constructor/50885340#50885340 
Maybe there is another way?

**Comment by ThomasKirst** [comment, marked as answer]
Hi Alex,
This is a very specific problem. The question is: do you really need a class here?
In a simple implementation you would create a function which returns an object which is of type `Artist`.

The function would look like:
```js
async function createArtist(
    artistName: string,
    city: string,
    streetname: string,
    number: number,
    tatoos: string[]
  ): Promise<Artist> {
  // ... create the object, fetch geoData and return the full object
}
```

When using classes and a constructor it seems very difficult with the suggested approach mentioned on StackOverflow (making the constructor an immediately invoked async function).

An approach in Object Oriented Programming is generally using a so called 'Factory Function' which creates the object (and maybe it does some checks / validation upfront).

You would make the constructor `private`. Why this might be a good idea is explained in this article, for instance:
–> https://khalilstemmler.com/blogs/typescript/when-to-use-a-private-constructor/

A static `Artist.create()` function would then instantiate the Artist and perform the async operation to fetch the geo data.

Thereafter the `create()` function will return the fully populated object.

The call to `Artist.create()` would then be:
```js
const artist = await Artist.create('TK ONE', 'Munich', 'Highway', 1, []);
/*
    artistName: "TK ONE"
    location: Object
      city: "Munich"
      streetname: "Highway"
      number: 1
    slug: "tkone"
    tattoos: Array(0)
    position: Array(2)
      0: 0
      1: 0
*/
```

I created a CodeSandbox where you can have a look at this possible solution:
–> https://codesandbox.io/s/object-instantiation-with-a-factory-method-4jrygx

**Comment by onebarloop** [comment]
Thomas, thank you so much for your time. My initial motivation for using a class-constructor was to use the "normalize"-function (which depends on a prop of the object, i.e. "artistName") during the creation of the artist-object. 

This is a very interesting article, I will try to apply your solution!

**Comment by onebarloop** [comment]
Worked like a charm: https://github.com/onebarloop/capstone-project/pull/33
